vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,434 @@
+      }
+
+      void await_resume() const noexcept {}
+
+      void await_suspend(std::experimental::coroutine_handle<> resume)
+      {
+        m_queue.Post([context = resume.address()]() noexcept
+        {
+          std::experimental::coroutine_handle<>::from_address(context)();
+        });
+      }
+
+    private:
+      Mso::DispatchQueue m_queue;
+    };
+
+    return awaitable{ queue };
+  }
+}
+
+namespace Microsoft::React
+{
+WinRTWebSocketResource::WinRTWebSocketResource(Uri&& uri, vector<ChainValidationResult> certExeptions)
+  : m_uri{ std::move(uri) }
+{
+  m_socket.MessageReceived({ this, &WinRTWebSocketResource::OnMessageReceived });
+
+  for (auto certException : certExeptions)
+  {
+    m_socket.Control().IgnorableServerCertificateErrors().Append(certException);
+  }
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString, vector<ChainValidationResult> certExeptions)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) }, certExeptions)
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Close(CloseCode::GoingAway, "Disposed");
+
+  m_closePerformed.Wait();
+}
+
+#pragma region Private members
+
+IAsyncAction WinRTWebSocketResource::PerformConnect()
+{
+  auto self = shared_from_this();
+  try
+  {
+    co_await resume_background();
+
+    co_await self->m_socket.ConnectAsync(self->m_uri);
+
+    self->m_readyState = ReadyState::Open;
+    if (self->m_connectHandler)
+    {
+      self->m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    self->m_readyState = ReadyState::Closed;
+    if (self->m_errorHandler)
+    {
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+    }
+  }
+
+  SetEvent(self->m_connectPerformed.get());
+  self->m_connectPerformedPromise.set_value();
+  self->m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
+{
+  auto self = shared_from_this();
+  try
+  {
+    co_await resume_background();
+
+    co_await resume_on_signal(self->m_connectPerformed.get());
+
+    if (self->m_readyState != ReadyState::Open)
+    {
+      self = nullptr;
+      co_return;
+    }
+
+    self->m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    self->m_writer.WriteBytes(arr);
+
+    co_await self->m_writer.StoreAsync();
+
+    if (self->m_pingHandler)
+    {
+      self->m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    if (self->m_errorHandler)
+    {
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  auto self = shared_from_this();
+  if (self->m_writeQueue.empty())
+  {
+    self = nullptr;
+    co_return;
+  }
+
+  try
+  {
+    co_await resume_background();
```

Comment:
I've been wondering. If network operations are already performed asynchronously, why are we shuttling work to the background threadpool? I.e. what expensive operations are we wanting to avoid on the main thread?

Reply:
Connection is a good example.
In the past, we did see noticeable in the consuming apps when the connection was handled in the main thread.

Reads and writes can potentially be heavier, depending on the message's size.